File: backend/infraction-service/app/com/radovan/play/filters/UnifiedMetricsFilter.scala
package com.radovan.play.filters

import com.radovan.play.services.PrometheusService
import jakarta.inject.{Inject, Singleton}
import play.api.mvc.{EssentialAction, EssentialFilter}
import scala.concurrent.ExecutionContext

@Singleton
class UnifiedMetricsFilter @Inject()(
                                      prometheus: PrometheusService
                                    )(implicit ec: ExecutionContext) extends EssentialFilter {

  override def apply(next: EssentialAction): EssentialAction = EssentialAction { requestHeader =>
    // 🚫 Preskoči metrikovanje za Prometheus scrape i Health check
    if (requestHeader.path == "/prometheus" || requestHeader.path == "/api/health") {
      next(requestHeader)
    } else {
      val startTimeNs = System.nanoTime()
      prometheus.increaseRequestCount()

      next(requestHeader).map { result =>
        val durationSec = (System.nanoTime() - startTimeNs) / 1e9
        prometheus.recordResponseTime(durationSec)
        prometheus.updateHttpStatusCount(result.header.status)
        result
      }
    }
  }
}

File: backend/registration-service/app/com/radovan/play/filters/UnifiedMetricsFilter.java
package com.radovan.play.filters;

import com.radovan.play.services.PrometheusService;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import org.apache.pekko.util.ByteString;
import play.libs.streams.Accumulator;
import play.mvc.EssentialAction;
import play.mvc.EssentialFilter;
import play.mvc.Result;

import java.util.List;
import java.util.concurrent.Executor;

@Singleton
public class UnifiedMetricsFilter extends EssentialFilter {

    private PrometheusService prometheus;
    private Executor executor;

    // Lista endpoint-a koji treba da budu izuzeti iz metrika
    private final List<String> excludedPaths = List.of(
            "/prometheus","/api/health"
    );

    @Inject
    public void initialize(PrometheusService prometheus, Executor executor) {
        this.prometheus = prometheus;
        this.executor = executor;
    }

    @Override
    public EssentialAction apply(EssentialAction next) {
        return EssentialAction.of(request -> {
            // Proveri da li je putanja izuzeta
            if (isExcludedPath(request.path())) {
                return next.apply(request); // Preskoči metrike za ovaj endpoint
            }

            long startTimeNs = System.nanoTime();

            // 📈 Increase request count immediately
            prometheus.increaseRequestCount();

            Accumulator<ByteString, Result> accumulator = next.apply(request);

            return accumulator.map(result -> {
                double durationSec = (System.nanoTime() - startTimeNs) / 1_000_000_000.0;

                // ⏱ Record response time
                prometheus.recordResponseTime(durationSec);

                // 📊 Status classification (2xx, 4xx, 5xx)
                prometheus.updateHttpStatusCount(result.status());

                return result;
            }, executor);
        });
    }

    /**
     * Proverava da li putanja treba da bude izuzeta iz metrika
     */
    private boolean isExcludedPath(String path) {
        return excludedPaths.stream()
                .anyMatch(excludedPath ->
                        path.equals(excludedPath) ||
                                path.startsWith(excludedPath + "/"));
    }
}